### PR TITLE
[fixes #4495] - Quarkus Keycloak Authorization Extension

### LINKED ISSUE #4495: OIDC adapter has to provide HttpPermissionChecker
**Description**
Example of the original `quarkus-keycloak` configuration:
```
quarkus.keycloak.policy-enforcer.claim-information-point.claims.request-uri={request.relativePath}
quarkus.keycloak.policy-enforcer.claim-information-point.claims.request-method={request.method}
quarkus.keycloak.policy-enforcer.lazy-load-paths=true

quarkus.keycloak.policy-enforcer.paths.1.path=/asset-service/resources/tenants
quarkus.keycloak.policy-enforcer.paths.1.methods.1.method=POST
quarkus.keycloak.policy-enforcer.paths.1.methods.1.scopes=tenants:create

quarkus.keycloak.policy-enforcer.paths.10.path=/asset-service/resources/tenants/{tenantId}
quarkus.keycloak.policy-enforcer.paths.10.methods.1.method=PUT
quarkus.keycloak.policy-enforcer.paths.10.methods.1.scopes=tenant:edit
quarkus.keycloak.policy-enforcer.paths.10.methods.2.method=DELETE
quarkus.keycloak.policy-enforcer.paths.10.methods.2.scopes=tenant:delete
quarkus.keycloak.policy-enforcer.paths.11.path=/asset-service/resources/tenants/{tenantId}/disable
quarkus.keycloak.policy-enforcer.paths.11.methods.1.method=PUT
quarkus.keycloak.policy-enforcer.paths.11.methods.1.scopes=tenant:disable
quarkus.keycloak.policy-enforcer.paths.12.path=/asset-service/resources/tenants/{tenantId}/enable
quarkus.keycloak.policy-enforcer.paths.12.methods.1.method=PUT
quarkus.keycloak.policy-enforcer.paths.12.methods.1.scopes=tenant:enable
```
@dfranssen:

E.g. Keycloak javascript permission using the request-uri. Our path is like /asset-service/resources/tenants/abc where abc is the tenant name (uriPart[4] in the code below) for which we check the actual authenticated user is part of. In Keycloak we have a group per tenant (e.g. /abc/users) where the user is put into.
```
var attributes = context.getAttributes();
var httpUri = attributes.getValue('request-uri');
function isInTenantGroup(tenantId) {
    var result = false;
    var wanted = "/" + tenantId + "/users";
    var identity = context.getIdentity();
    var userGroups = identity.getAttributes().getValue('user-groups');
    if (userGroups !== null) {
        for(var i=0; i < userGroups.size(); i++) {
            if(wanted === userGroups.asString(i)) {
                result = true;
                break;
            }
        }
    }
    return result;
}

if (httpUri) {
    var uriParts = httpUri.asString(0).split('/');
    if (isInTenantGroup(uriParts[4])) {
        $evaluation.grant();
    }
}
```

**Implementation ideas**
The adapter has to introspect a `scope` claim (or Keycloak specific claim - in this case we should have a claim name configured to avoid tying the adapter to KC -  though we can do it later). PolicyEnforcer configuration group can be introduced if needed

### PR DESCRIPTION
**Pull request by pedroigor**
Quickstart: https://github.com/quarkusio/quarkus-quickstarts/pull/331

### CODE REVIEW

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerAuthorizer.java
Comment (stuartwdouglas) [inline extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerAuthorizer.java]: So this is not really integrating with our HTTP permission layer, but just completely replacing it?

We need to discuss this, as it means that all other security rules will be ignored.

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerAuthorizer.java
Comment (stuartwdouglas) [inline extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerAuthorizer.java]: This is a blocking operation, we need to make sure we are not on an IO thread. This API will need to be changed to add an easy way to run blocking operations.

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java
Comment (stuartwdouglas) [inline extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java]: Is there any reason that this can't use the existing path stuff from our current authorization layer? The original idea was that the policies would be pluggable, and there would be one place to map paths to policies.

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerAuthorizer.java
Comment (stuartwdouglas) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerAuthorizer.java]: Actually I think this needs two layers, the existing policy based path stuff, and the ability to add additional checks that do their own path matching (which servlet web.xml integration will need).

**Comment by sberyozkin** [comment]
@pedroigor thanks, I'll be commenting as well during the next few days. One thing is that we indeed should try to reuse as much as possible from the lower levels like the core security and oidc, and if it is not possible, push down some changes. (re the comment from @stuartwdouglas about the paths for ex)

File: docs/src/main/asciidoc/keycloak-authorization-guide.adoc
Comment (sberyozkin) [inline docs/src/main/asciidoc/keycloak-authorization-guide.adoc]: Is this `secret` actually used by the enforcer ? Just checking, it was not used in the `integration-tests/oidc`, but may be the enforcer does need it

File: extensions/keycloak-authorization/deployment/pom.xml
Comment (sberyozkin) [inline extensions/keycloak-authorization/deployment/pom.xml]: This should probably be `Quarkus - Keycloak Authorization - Deployment` (same for the /runtime) but the description can indeed be more specific

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerAuthorizer.java
Comment (sberyozkin) [inline extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerAuthorizer.java]: There probably should be some abstract code based on what Stuart @stuartwdouglas did, so that the default authorizer and this one do not duplicate. 
What was the problem with only plugging in a custom `HttpSecurityPolicy` ?

File: extensions/keycloak-authorization/runtime/src/main/resources/META-INF/quarkus-extension.json
Comment (gsmet) [inline extensions/keycloak-authorization/runtime/src/main/resources/META-INF/quarkus-extension.json]: Please also add it in the legacy json for now: ./devtools/platform-descriptor-legacy/src/main/filtered/extensions.json

**Review by gsmet — CHANGES_REQUESTED** [review]
I reviewed the doc and added some suggestions.

Question is: should this extension go in if it doesn't support native? Until now, it was a strict condition.

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java
Comment (sberyozkin) [inline extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java]: I believe this has to be managed by the `quarkus-oidc` - this is exactly what it will already do as part of your authorization code flow :-) or the existing bearer-only check. I mean it will do 401 if the token is invalid (signature for ex), but if the token has not enough permissions, it is equally invalid. And we let the user be redirected to something local then we'll lose the flow control, what will happen after the user lands on some local page, as opposed to be redirected to IDP ?

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java
Comment (sberyozkin) [inline extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java]: IMHO everything not related to the Keycloak specific policy-enforcer configuration should not be here at all; we've promised `quarkus-oidc` will work with not only KC :-), and UMA is supported not only by Keycloak

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java
Comment (sberyozkin) [inline extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java]: I'm not sure how many of those properties are related to the KC policy enforcer configuration the user was trying to migrate. Please review and leave only those which are needed to support that policy-enforcer configuration; everything else can go later on to `quarkus-oidc` or `quarkus-oidc-uma` for example

File: integration-tests/keycloak-authorization/src/main/resources/keycloak.json
Comment (sberyozkin) [inline integration-tests/keycloak-authorization/src/main/resources/keycloak.json]: where is this `keycloak.json` resource used ?

File: extensions/keycloak-authorization/runtime/src/main/resources/META-INF/quarkus-extension.json
Comment (pedroigor) [reply extensions/keycloak-authorization/runtime/src/main/resources/META-INF/quarkus-extension.json]: It is there already.

**Comment by pedroigor** [comment]
> I reviewed the doc and added some suggestions.
> 
> Question is: should this extension go in if it doesn't support native? Until now, it was a strict condition.

I can work on native and it should not take too much time. We had native work for the old Keycloak extension before so is just a matter of doing the same changes.

@sberyozkin suggested leaving native for now though. But I can deliver it too.

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerAuthorizer.java
Comment (pedroigor) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerAuthorizer.java]: @stuartwdouglas this extension is kind of a replacement of Quarkus permissions. Differently than what we have in Quarkus, the policies are externalized, completely decoupled from the application, and centralized managed through Keycloak.

The functionality provided by our policy enforcer has more requirements in regards to path mapping and matching. We also have some requirements in regards to not authorized or forbidden responses (e.g.: when doing UMA).

This extension is all about bringing a Keycloak specific feature into Quarkus. We have users familiar with the configuration I'm proposing here, making easier to migrate their applications.

I'm not sure if it would make sense to have part of your resources protected by Quarkus policies and other parts protected by Keycloak. However, we could review the changes I'm proposing and delegate to the Quarkus permission layer when the extension denies access to a resource.

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java
Comment (pedroigor) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java]: The key point here is that we are not just plugging in a new policy, but a completely different solution to authorization.

I tried to do it using what Quarkus provides, but that would mean bringing too many things that are really specific to this feature in Keycloak.

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerAuthorizer.java
Comment (pedroigor) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerAuthorizer.java]: I don't think there is a duplication. The authorization process is different if you compare both solutions.

This extension just brings an existing and specific feature from Keycloak into Quarkus. It is related to centralized authorization and how to externalize authorization from applications based on a Resource-Based Access Control.

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java
Comment (pedroigor) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java]: This is not really related to code flow. This option, in particular, is just providing an option to customize requests denied by the server. The user is already authenticated and if not, this page is not meant to start any flow.

For instance, instead of returning a 403 response the application may want to show a more fancy page.

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java
Comment (pedroigor) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java]: I see. So we would be providing a separated extension for UMA support if someone demands it ?

Sounds good for me. I can remove UMA support from this extension.

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java
Comment (pedroigor) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java]: Except for the `userManagedAccess` all other properties are specific to KC policy enforcer.

**Comment by pedroigor** [comment]
@gsmet Native support added.

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerAuthorizer.java
Comment (pedroigor) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerAuthorizer.java]: Can we follow the same approach as you did to the `IdentityProviderManager`? So that we pass an object to authorizers similar to `AuthenticationRequestContext` (maybe rename it to become more generic) in order to run blocking tasks?

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java
Comment (sberyozkin) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java]: @pedroigor hey Pedro, exactly, lets have it done either directly as part of `quarkus-oidc` (it is nor really that fat so may be it should be there) or a new extension like `quarkus-oidc-uma` - so that we can say Quarkus can be UMA protected with any IDP which does it, which will be massive. Let me open an issue to track it, it would probably be a copy and paste for you :-), may be the only minor effort would be required to replace the KC adapter client code with some basic HTTP client....

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java
Comment (sberyozkin) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java]: @pedroigor It looks like the error can only occur in the process of the user authenticating (for the bearer only flow it is only 403 anyway). So if it is SPA then it is only really an adapter level issue what happens to a user view. And if it is a server web app code flow then the user has signed in at the IDP site, is being redirected now to the adapter, the token fails a check, and with custom page the user lands there seeing that the access is denied, But what is next ? The user stays authenticated ? Does not seem like the right idea.
The other thing, I doubt that in production people will want to display the details outside of its log :-)

Perhaps we should have an option for OIDC adapter itself return a page like "Your access to this resource has been denied, please contact the administrator, you will be redirected to IDO in 5 sec..." (and this page can be customized). 

But whatever we do here, it feels like it is not a KC specific extension concern since 403 will be possible at the `quarkus-oidc` level as well :-)

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerAuthorizer.java
Comment (sberyozkin) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerAuthorizer.java]: @pedroigor hi, so given we've agreed `UMA` will be managed differently, how will that be done, will it have its own authorization mechanism as well ? Should we give it a try and have the core authorization mechanism be pluggable/customizabl as needed to meet KV ploicy-enforcer requirements ?
I have to admit I don't quite understand, if you could explain what is too rigid/etc in the core one ? Sorry if I'm missing something...

**Comment by sberyozkin** [comment]
@pedroigor @gsmet yes, since Pedro was having some issues I suggested leaving it for now in case if would block the PR, the user who can't migrate may not be using the native mode...Nice to see the native is supported now after all :-)

**Comment by pedroigor** [comment]
@sberyozkin Yeah. I was missing something when facing that error. Sorry for the false alarm :)

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java
Comment (pedroigor) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java]: Client-side, UMA is quite straight forward to implement. I agree that a separated extension is the best way.

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java
Comment (pedroigor) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java]: That is the idea. This requirement came from the community a long time ago.

The user would just see this custom page with whatever the application wants to show. Either a simple and fancy "Access Denied" or something else.

If we want to have that on `quarkus-oidc`, that is fine and I can remove this.

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerAuthorizer.java
Comment (pedroigor) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerAuthorizer.java]: The core one is not rigid, on the contrary, it is quite flexible. I just couldn't find a nice way to plug the policy enforcer capabilities. 

However, that is not Quarkus fault. But the nature of this extension that aims a different authorization model.

If you look at the guide, for instance, you'll see that users can enable the policy enforcer without even specify a single path. Where the mapping between paths and policies are done by the extension itself, through some communication with the Keycloak server.

I see this extension as similar to any other that brings functionality from a 3rd party framework/library. We are also constantly involving the capabilities, adding new ones, etc. We may even end up in a situation where a new capability would require changes in Quarkus.

So, what Stuart did is flexible enough from both a user perspective in regards to declarative authorization (support for different/custom policies, reuse policies across different permissions, etc) as well as from a developer perspective, like in this case, where we are enabling centralized authorization for Quarkus apps.

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java
Comment (stuartwdouglas) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java]: I won't hold up this PR for this, but we should create an issue to track it.

**Comment by stuartwdouglas** [comment]
I have pushed an extra commit onto this that will run the native tests in CI. If this passes then I think we are good to go.

**Review by stuartwdouglas — APPROVED** [review]

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java
Comment (sberyozkin) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java]: @pedroigor yes please lets remove if for the moment and then figure out where to have this property - the same concept should work across the OIDC bits in Quarkus

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java
Comment (sberyozkin) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/KeycloakPolicyEnforcerConfig.java]: @stuartwdouglas we get it in too fast and the users will see the extra properties in the wrong namespace and a new migration headache is on the way

**Comment by sberyozkin** [comment]
@stuartwdouglas we don't have to rush. It is nearly ready but UMa does not belong at the KC level, similarly to the redirect page property - these are the extra bits which are not required for the KC policy-enforcer migration support. IMHO the consolidation around the quarkus-oidc is the key

**Review by sberyozkin — CHANGES_REQUESTED** [review]
Hi Pedro, Stuart, @pedroigor @stuartwdouglas I'm only asking here not for some extra improvements, but only to limit what is currently exposed at the quarkus-keycloak-authorization configuration level, specifically `UMA-related and redirect related` configs.
My point really is we've gone for a major upgrade of the OIDC code, dropped the original keycloak extension, hence IMHO we should be trying to keep all what can be generic at the oidc level to avoid people migrating again when we will support oidc-uma, etc. If we get these properties going into 0.27.0 then it is a new concern. Hope you are all OK with it.

**Comment by pedroigor** [comment]
@sberyozkin Sure. Both options are not available anymore.

**Comment by sberyozkin** [comment]
@pedroigor thanks :-) and we'll get them reborn pretty soon for sure :-)

File: integration-tests/keycloak-authorization/src/main/java/io/quarkus/it/keycloak/ProtectedResource.java
Comment (sberyozkin) [inline integration-tests/keycloak-authorization/src/main/java/io/quarkus/it/keycloak/ProtectedResource.java]: @pedroigor I'll add a follow up issue for us to come up with Quarkus permission class so that the users can write a portable code with or without KC policy enforcer, not a prob for this PR 
CC @stuartwdouglas

**Review by sberyozkin — APPROVED** [review]

**Pull request merged**
